File: src/lib/models/check-result.ts
import mongoose, { Schema, type Document } from "mongoose";

export interface ICheckResult extends Document {
  endpointId: mongoose.Types.ObjectId;
  status: "UP" | "DOWN" | "DEGRADED";
  statusCode: number | null;
  responseTime: number | null;
  error: string | null;
  checkedAt: Date;
}

const DATA_RETENTION_DAYS = 30;

const CheckResultSchema = new Schema<ICheckResult>({
  endpointId: {
    type: Schema.Types.ObjectId,
    ref: "Endpoint",
    required: true,
  },
  status: {
    type: String,
    enum: ["UP", "DOWN", "DEGRADED"],
    required: true,
  },
  statusCode: { type: Number, default: null },
  responseTime: { type: Number, default: null },
  error: { type: String, default: null },
  checkedAt: { type: Date, default: Date.now },
});

CheckResultSchema.index({ endpointId: 1, checkedAt: -1 });
CheckResultSchema.index(
  { checkedAt: 1 },
  { expireAfterSeconds: DATA_RETENTION_DAYS * 24 * 60 * 60 }
);

export default mongoose.models.CheckResult ||
  mongoose.model<ICheckResult>("CheckResult", CheckResultSchema);

File: backend/models/check_result.py
from datetime import datetime, timezone
from typing import Literal, Optional

from beanie import Document, PydanticObjectId
from pydantic import Field
from pymongo import IndexModel, ASCENDING, DESCENDING

from core.config import settings


class CheckResult(Document):
    endpoint_id: PydanticObjectId = Field(alias="endpointId")
    status: Literal["UP", "DOWN", "DEGRADED"]
    status_code: Optional[int] = Field(default=None, alias="statusCode")
    response_time: Optional[float] = Field(default=None, alias="responseTime")
    error: Optional[str] = None
    checked_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), alias="checkedAt")

    class Settings:
        name = "check_results"
        indexes = [
            IndexModel(
                [("endpointId", ASCENDING), ("checkedAt", DESCENDING)],
            ),
            IndexModel(
                [("checkedAt", ASCENDING)],
                expireAfterSeconds=settings.DATA_RETENTION_DAYS * 86400,
            ),
        ]

    model_config = {"populate_by_name": True}
